### PR TITLE
Remove decorators and use webpack to build {N} app with AoT

### PR DESCRIPTION
**Pull request by NathanWalker**
@m-abs @d3viant0ne let's continue here.
Working webpack build, just a runtime error:
```
file:///app/bundle.js:37207:49: JS ERROR TypeError: moduleFactory.create is not a function. (In 'moduleFactory.create(ngZoneInjector)', 'moduleFactory.create' is undefined)
```

### CODE REVIEW

File: nativescript/webpack.common.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+          exclude: /app\.css$/,
+          use: [
+            'raw-loader',
+          ]
+        },
+
+        // Compile TypeScript files with ahead-of-time compiler.
+        {
+          test: /\.ts$/,
+          loaders: [
```
Comment (m-abs) [inline nativescript/webpack.common.js]: @NathanWalker 
Why did you change all but this one from `loaders` to `use`?

File: nativescript/webpack.common.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      }),
+      //Generate a bundle starter script and activate it in package.json
+      new nsWebpack.GenerateBundleStarterPlugin([
+        './vendor',
+        './bundle',
+      ]),
+      //Angular AOT compiler
+      new AotPlugin({
+        tsConfigPath: 'tsconfig.aot.json',
+        entryModule: 'app/native.module#NativeModule',
```
Comment (m-abs) [inline nativescript/webpack.common.js]: I'm wondering if the *.aot.ts should have the same basename as entryModule?
The current aot.ts-file is named app.aot.ts, maybe it should be named native.module.aot.ts?

File: nativescript/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        }
+      },
+      {
+        from: {
+          glob: '**/*.component.*.' + process.env.PLATFORM,
+          dot: true
+        },
+        to: '[path]/[name]'
+      },
+      // @nota plugins
```
Comment (m-abs) [inline nativescript/webpack.config.js]: Wupsy. this shouldn't have been added to the seed project. :-1: 
This is one of the modules, we use in LYT3, the project I'm working on.

File: nativescript/package.json [hunk 4]
```diff
@@ -1,58 +1,83 @@
...
-  "version": "0.0.0",
-  "nativescript": {
-    "id": "com.yourdomain.appname",
-    "tns-ios": {
-      "version": "2.4.0"
+    "name": "angular-seed-advanced",
+    "main": ".tmp/app.js",
+    "version": "0.0.0",
+    "nativescript": {
+        "id": "com.yourdomain.nativescript",
```
Comment (m-abs) [inline nativescript/package.json:6]: This doesn't match the name in the `app.gradle`.

File: tools/config/seed-advanced.config.ts [hunk 5]
```diff
@@ -1,11 +1,30 @@
...
+
+  TNS_APP_DEST = `${this.TNS_BASE_DIR}/${this.destSubdir}`;
+
+  TNS_CONFIG = {
+    ANALYTICS_TRACKING_ID: '',
+  };
 
   constructor() {
     super();
+    this.ENABLE_SCSS = true;
```
Comment (m-abs) [inline tools/config/seed-advanced.config.ts:27]: Should SCSS be enabled? I think this line got copied by accident from our project.

File: src/client/app/frameworks/sample/sample.module.ts [hunk 6]
```diff
@@ -28,6 +28,10 @@ import { MultilingualModule } from '../i18n/multilingual.module';
   providers: [
     SAMPLE_PROVIDERS
   ],
+  schemas: [
+    NO_ERRORS_SCHEMA,
+    CUSTOM_ELEMENTS_SCHEMA
```
Comment (m-abs) [inline src/client/app/frameworks/sample/sample.module.ts:33]: I understand `NO_ERRORS_SCHEMA`, but why `CUSTOM_ELEMENTS_SCHEMA`?

File: src/client/app/frameworks/core/services/app.service.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { Injectable } from '@angular/core';
+
+// app
+import { AnalyticsService } from '../../../frameworks/analytics/index';
+import { LogService } from './log.service';
+import { Config } from '../utils/config';
+import { MultilingualService } from '../../../frameworks/i18n/index';
+
+@Injectable()
+export class AppService {
```
Comment (m-abs) [inline src/client/app/frameworks/core/services/app.service.ts]: Just to make it clear, since I haven't explained it.

`AppService` and `NSAppService` were created because `nativescript-dev-webpack` didn't like the `NSAppComponent` that inherited from `AppComponent` but wasn't real a `@Component`.
It's only supported to get injected in the `AppComponent`, the idea is to setup `NativeScript` listeners and more.

File: src/client/app/components/home/home.component.ts [hunk 8]
```diff
@@ -1,21 +1,28 @@
...
   templateUrl: 'home.component.html',
-  styleUrls: ['home.component.css']
+  styleUrls: [
+    'home.component.css',
+  ],
 })
 export class HomeComponent {
   public names$: Observable<any>;
   public newName: string = '';
+  private _cnt: number = 3; // just for example purposes
```
Comment (m-abs) [inline src/client/app/components/home/home.component.ts]: Leftover from something? It's not used.

File: nativescript/src/shared/core/services/ns-app.service.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+import * as multilingual from '../../../app/frameworks/i18n/index';
+
+declare var android: any;
+const trackingId = '<%= TNS_CONFIG.ANALYTICS_TRACKING_ID %>';
+
+@Injectable()
+export class NSAppService extends AppService {
+  nsApp = nsApp;
+
+  externalRouteMap = [
```
Comment (m-abs) [inline nativescript/src/shared/core/services/ns-app.service.ts]: Wupsy: More leftover from our project.
This is meant for deeplink we use.

File: nativescript/src/_common.scss [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+$ab-color: $white;
+$btn-color: $accent;
+$btn-color-secondary: darken($accent, 5%);
+
+// Import the theme’s main ruleset.
+@import 'nativescript-theme-core/scss/index';
+
+// TextField {
+//   margin: 15 15 0 15;
+// }
```
Comment (joshwiens) [inline nativescript/src/_common.scss]: Pull the commented out block

File: nativescript/src/components.module.ts [hunk 11]
```diff
@@ -4,16 +4,16 @@ import { NativeScriptModule, NativeScriptFormsModule, NativeScriptHttpModule, Na
...
 
 // libs
 import { StoreModule } from '@ngrx/store';
 import { EffectsModule } from '@ngrx/effects';
 import { TranslateModule, TranslateLoader, TranslateStaticLoader } from 'ng2-translate';
 
 // app
-import { WindowService, ConsoleService, RouterExtensions } from './app/frameworks/core/index';
-import { NSAppComponent } from './pages/app/app.component';
+import { WindowService, ConsoleService, RouterExtensions, AppService } from './app/frameworks/core/index';
```
Comment (joshwiens) [inline nativescript/src/components.module.ts]:
```
import { 
  WindowService, 
  ConsoleService, 
  RouterExtensions, 
  AppService 
} from './app/frameworks/core/index';
```

For anything with more than three imports, particularly with something with a long path multi-line imports tend to be a bit more readable imo.

File: nativescript/src/native.module.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
+// nativescript
+import { NativeScriptModule, NativeScriptFormsModule, NativeScriptHttpModule, NativeScriptRouterModule, RouterExtensions as TNSRouterExtensions } from 'nativescript-angular';
```
Comment (joshwiens) [inline nativescript/src/native.module.ts]: Same as above, I would multi-line this one as well

File: nativescript/src/tns-java-classes.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
+//Resolve JavaScript classes that extend a Java class, and need to resolve
+//their JavaScript module from a bundled script. For example:
+//NativeScriptApplication, NativeScriptActivity, etc.
+//
+//This module gets bundled together with the rest of the app code and the
+//`require` calls get resolved to the correct bundling import call.
+//
+//At runtime the module gets loaded *before* the rest of the app code, so code
+//placed here needs to be careful about its dependencies.
```
Comment (joshwiens) [inline nativescript/src/tns-java-classes.ts]: Would multi-line jsdoc style comments be a bit more readable than 9 single line comments?

File: nativescript/src/app.aot.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
+// this import should be first in order to load some required settings (like globals and reflect-metadata)
+import { platformNativeScript } from "nativescript-angular/platform-static";
```
Comment (joshwiens) [inline nativescript/src/app.aot.ts]: This will fail with the current tslint settings

```
    "quotemark": [
      true,
      "single"
    ],
```

File: nativescript/src/tns-java-classes.ts [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    global.__requireOverride = function (name, dir) {
+        if (name === "./tns_modules/application/application.js") {
+            return require("application");
+        } else if (name === "./tns_modules/ui/frame/frame.js") {
+            return require("ui/frame");
+        } else if (name === "./tns_modules/ui/frame/activity.js") {
+            return require("ui/frame/activity");
+        }
+    };
+}
```
Comment (joshwiens) [inline nativescript/src/tns-java-classes.ts]: This will fail with the current tslint settings

```
    "quotemark": [
      true,
      "single"
    ],
```

File: nativescript/src/vendor-platform.android.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    global.__requireOverride = function (name, dir) {
+        if (name === "./tns_modules/application/application.js") {
+            return require("application");
+        } else if (name === "./tns_modules/ui/frame/frame.js") {
+            return require("ui/frame");
+        } else if (name === "./tns_modules/ui/frame/activity.js") {
+            return require("ui/frame/activity");
+        }
+    };
+}
```
Comment (joshwiens) [inline nativescript/src/vendor-platform.android.ts:25]: This will fail with the current tslint settings

```
    "quotemark": [
      true,
      "single"
    ],
```

File: tools/tasks/seed/build.js.tns.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+        //    .pipe(plugins.sourcemaps.write('.', {
+        //      includeContent: false,
+        //      sourceRoot: (file: any) =>
+        //        relative(file.path, PROJECT_ROOT + '/' + APP_SRC).replace(sep, '/') + '/' + APP_SRC
+        //    }))
+        .pipe(plugins.template(template))
+        .pipe(gulp.dest(Config.TNS_APP_DEST));
+
+      const copy = projectFiles()
+        //.pipe(plugins.template(template))
```
Comment (joshwiens) [inline tools/tasks/seed/build.js.tns.ts]: Should be pulled if we aren't using it.

File: nativescript/src/tns-java-classes.ts (hunk 13 above)
Comment (m-abs) [reply nativescript/src/tns-java-classes.ts]: This file should be removed, it's from the old-version of `nativescript-dev-webpack`

File: tools/tasks/seed/build.js.tns.ts (hunk 17 above)
Comment (m-abs) [reply tools/tasks/seed/build.js.tns.ts]: It was commented out because the template plugin threw an error because of it.
It should either get fixed or better yet implemented in webpack.

**Comment by m-abs** [comment]
Could we do something about node-sass that keeps failing in ci?

File: nativescript/src/tns-java-classes.ts (hunk 13 above)
Comment (m-abs) [reply nativescript/src/tns-java-classes.ts]: In newer versions the same functionality is covered by vendor.ts and vendor-platform.android/ios.ts, they look just as bad, but they're from a template so I don't think we should alter them too much.

File: src/client/app/frameworks/sample/sample.module.ts (hunk 6 above)
Comment (NathanWalker) [reply src/client/app/frameworks/sample/sample.module.ts:33]: http://stackoverflow.com/questions/39428132/custom-elements-schema-added-to-ngmodule-schemas-still-showing-error

File: nativescript/package.json (hunk 4 above)
Comment (NathanWalker) [reply nativescript/package.json:6]: I have updated `app.gradle` to match.

File: nativescript/webpack.common.js (hunk 2 above)
Comment (NathanWalker) [reply nativescript/webpack.common.js]: I don't think that matters here as it's the app.ts file which is the `main` targeted from `package`.

File: nativescript/webpack.common.js (hunk 1 above)
Comment (NathanWalker) [reply nativescript/webpack.common.js]: thanks forgot that one, updated now.

**Comment by NathanWalker** [comment]
I'm not concerned about CI until we can at least get everything else working.

File: tools/tasks/seed/build.js.tns.ts (hunk 17 above)
Comment (joshwiens) [reply tools/tasks/seed/build.js.tns.ts]: I'd vote for just moving that into Webpack as well.

File: src/client/app/frameworks/core/services/app.service.ts (hunk 7 above)
Comment (NathanWalker) [reply src/client/app/frameworks/core/services/app.service.ts]: Love it, great idea 👍

**Comment by joshwiens** [comment]
But how do we know it's working without CI :p

**Comment by m-abs** [comment]
I discovered that I could break my hallo world project by upgrading to angular 2.2.3 from 2.2.1. So I downgraded this project to 2.2.1. I also discovered that I had a new version og webpack in this project than in my hallo world, so I downgraded that aswell.

Now I get a new Error when I try to run this project.

```
Cannot redefine property: Config in....
```

I'm looking into what happens here, but it looks like progress from the NativeModule is not an NgModule error I had before.

**Comment by m-abs** [comment]
Once I solved that problem, I got the NgModule error again...

**Comment by m-abs** [comment]
I fixed most of the errors I got from running ngc manually.
It complains about paths to includes in the generated ngfactory.ts-files, but I think this might be the reason for the webpack-plugin?
```
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/app/components/about/about.component.ngfactory.ts:22:27: Cannot find module '../../../../ui/page'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/app/components/app.component.ngfactory.ts:31:27: Cannot find module '../../../ui/page'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/app/components/app.component.ngfactory.ts:39:27: Cannot find module '../../../ui/frame'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:69:27: Cannot find module '../ui/frame'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:76:27: Cannot find module '../platform'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:103:27: Cannot find module '../ui/page'.
Error at /home/mabs/project/angular2-seed-advanced/nativescript/app/native.module.ngfactory.ts:104:27: Cannot find module '../src/translate.service'.
```

The gulp build.tns-task didn't copy the aot.ts file into the app folder, so webpack didn't attempt to build it...
Unfortunately I now get this error from webpack:
```
ERROR in ./app.aot.ts                                                                                                                                                                                                                  
 Module build failed: Error: /home/mabs/project/angular2-seed-advanced/nativescript/app/app.aot.ts (4,39): Cannot find module './native.module.ngfactory'.)                                                                                
     at _transpile (/home/mabs/project/angular2-seed-advanced/nativescript/node_modules/@ngtools/webpack/src/loader.js:101:19)                                                                                                           
     at /home/mabs/project/angular2-seed-advanced/nativescript/node_modules/@ngtools/webpack/src/loader.js:128:26                                                                                                                        
     at process._tickCallback (internal/process/next_tick.js:103:7)                                                                                                                                                                      
```

**Comment by m-abs** [comment]
Hi @NathanWalker 

It took a lot of trail and error, but now I can both build and run with webpack bundles on this branch (at least on Android, I hope iOS works as well)

I had to move TranslateModule.forRoot(...) to `ComponentModule` from your `MultilingualModule` and I rolled `tns-core-modules` back to `2.4.3` from `next` and add `baseUrl` to the `tsconfig.aot.json`.
The problems I had building `app.aot.ts` were caused by strange import paths generated by `ngc`, especially for `ng2-translate`.

We still need one major thing before this can be merged. Template variables must be ported from the gulp tasks to this one.

I also stopped using the `tns-loader`, since it have no real function as long as the gulp task copies/renames the template files. I left in git, since we are going to need it.

It's my hope when `nativescript-dev-webpack` supports livesync, we'd no longer need the `src`-folder and we no longer need the gulp tns-tasks.

@NathanWalker could you merge master into this branch when you have some time for it? I could resolve the conflict or I might just be too tired. :)

**Comment by NathanWalker** [comment]
@m-abs that is very exciting news!

I will get latest merged up to this branch early next week, thanks 👍

**Comment by m-abs** [comment]
Hi again,

I'm getting some conflicts in the root project, rxjs was updated to @5.0.2 but both ng2-config and angular@2.2.3 wants a peerDep of rxjs@5.0.0_beta12.
*sic* Why are npm deps so troublesome?

I managed to fix most of the problems (so I at least could install root-project) by downgrading to typescript@2.1.4 (the first version to fix the symlink issue). I don't know what this will do to web aot builds...

**Comment by vip-git** [comment]
a quick question @NathanWalker and @m-abs :
- Webpack build is currently targeted towards nativescript or will this work on web build as well ?

I was planning to send a PR for webpack build for the webclient but if this is already taken care off then this is awesome !!

**Comment by m-abs** [comment]
@vip-git 
This PR only deals with nativescript.

@NathanWalker 
My last commit adds support for template variables in webpack and reenabled the `tns-loader`, which now have a purpose :)

I have to generate a JSON-file in the `gulp build.js.tns` task, otherwise I don't know how to extract the values from the `typescript`-files under `tools` and get them into webpack for tns.

Do you have a better suggestion? Something that'd work for both web/desktop and for tns?

**Comment by m-abs** [comment]
@NathanWalker 
Another issue (I keep running into new ones)
The `@Effects` decorator doesn't work with `AOT`.
Where is a workaround, which I'll look at tomorrow: https://github.com/ngrx/effects/pull/79

**Comment by NathanWalker** [comment]
@m-abs thank you for all the investigations. I must say `@Effects` decorator is a surprise, I would expect that to 'just work' but apparently everything with AoT is a chance...ergh.

**Comment by m-abs** [comment]
We have to wait until `nativescript-angular` have been updated to angular-2.4.x+, AOT and `@ngrx/effect` breaks with angular-2.3.x, right now the bundled App cannot start up due to a DI issue. 

I'll be working on other tasks for the next week or so.

**Comment by m-abs** [comment]
I've run into a new problem: https://github.com/NativeScript/nativescript-dev-webpack/issues/54

**Comment by m-abs** [comment]
I have bundling up and running now. We still need to test it properly.

**Comment by NathanWalker** [comment]
👏 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏  @m-abs bravo sir!
I'll take a look - would be amazing to get this merged this week. Thank you.

**Comment by NathanWalker** [comment]
@m-abs can you provide here the latest workflow with the latest changes?
For running development and when doing a release?

**Comment by m-abs** [comment]
***Development:***
`npm run start.livesync.android`
`npm run start.livesync.ios`

***Testing bundles:***
gulp build.prod.tns  && cd nativescript && npm run start-android-bundle
gulp build.prod.tns && cd nativescript && npm run start-ios-bundle
(Should be added as npm scripts)

***Publishing (Not tested):***
**iOS:**
gulp build.prod.tns --env-config prod --build-type prod && cd nativescript && tns publish ios --bundle --team-id <TEAM-ID>

**Android:**
gulp build.prod.tns --env-config prod --build-type prod && cd nativescript && tns build android --bundle --key-store-path <KEY-STORE-PATH> --key-store-password <KEY-STORE-PASS> --key-store-alias <KEY-STORE-ALIAS> --key-store-alias-password <KEY-STORE-ALIAS-PASSWORD>  --copy-to release.apk
Upload `release.apk` to Google Play.

In our project we have some scripts/tasks that handles building and publishing releases for both platforms.
I intend to clean these up and contribute them.

**Comment by m-abs** [comment]
@NathanWalker 
There is a few things we need before this is merged.

* Document: Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems.
* Document: Components/Directives/Pipes have to belong to a single `NgModule` or `ngc` will fail. This means an unused component have to be deleted and a component used in more than one `NgModule` have to be put in a single NgModule that the others can include.
* Document: The new dev, test, release workflow.
* Add: npm-scripts for building and releasing bundles.
* SCSS? Right now, because we use it in our project, the {N}-app uses SCSS but the angular-seed-project this is supposed to be optional.

**Comment by m-abs** [comment]
Then I tried to integrate this in our own project, ngc kept complaining about components/pipes inside `nativescript/src` not being in an NgModule, even though `nativescript/src` is in the `tsconfig.aot.json`'s exclude list.

I solved it by moving the directive in `CORE_DIRECTIVES` in to their own `NgModule` in `src/client/app/frameworks/core/directives/index.ts`.

I'm wondering if this is a good idea to do upstream?
Although I couldn't reproduce it with this project.

It seems to work fine with out own project, on my phone startup time have gone from 20+ seconds to ~2seconds.

**Comment by NathanWalker** [comment]
@m-abs 
* Let's enable SCSS by default (I'd consider that to be a desired out of box config for an advanced project like this) - I've always enabled on the 3 projects I have started myself based off this seed.
* Regarding directive in `CORE_DIRECTIVES` into own NgModule - *YES* let's move it here in the seed.
* `Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems.` - what are you referring to here?

**Comment by joshwiens** [comment]
`Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems` 

Curious about that myself

**Comment by m-abs** [comment]
> Service replaced with DI have to adhere to the API of the original class. This gave me a lot of problems

`NSAngulartics2Segment` is injected to replace `Angulartics2Segment` in {N}, which made `ngc` complain about the `NSAngulartics2Segment`.
If I remember correctly, `ngc` complained that `Angulartics2Segment` had a private `angulartics2`, while `NSAngulartics2Segment` didn't have one.
